### PR TITLE
Keep available manifest parsers in list, move config version parsing to separate function

### PR DESCRIPTION
**Pull request by ktrzcinx**
Improve code readability and allow to easily expand parsers list in future.

### CODE REVIEW

File: src/adsp_config.c [hunk 1]
```diff
@@ -1575,35 +1589,21 @@ static int adsp_parse_config_fd(FILE *fd, struct adsp *out, bool verbose)
...
-		ret = parse_adsp_config_v1_8(toml, out, verbose);
-	} else if (check_config_version(2, 5, manifest_version)) {
-		ret = parse_adsp_config_v2_5(toml, out, verbose);
-	} else {
-		ret = log_err(-EINVAL, "error: Unsupported config version %d.%d\n",
-			      manifest_version[0], manifest_version[1]);
-		goto error;
-	}
+	ret = log_err(-EINVAL, "error: Unsupported config version %d.%d\n",
+		      manifest_version[0], manifest_version[1]);
```
Comment (marc-hb) [inline src/adsp_config.c:1604]: This really obfuscates the `ret` value to save extremely little code.

File: src/adsp_config.c [hunk 2]
```diff
@@ -1575,35 +1589,21 @@ static int adsp_parse_config_fd(FILE *fd, struct adsp *out, bool verbose)
...
 
-	/* parse "version" array elements */
-	for (i = 0; i < ARRAY_SIZE(manifest_version); ++i) {
-		raw = toml_raw_at(arr, i);
-		if (ret < 0)
-			ret = err_key_parse("version", NULL);
-		ret = toml_rtoi(raw, &manifest_version[i]);
-		if (ret < 0)
-			ret = err_key_parse("version", "can't convert element to integer");
+	/* look for parser related with manifest version and run it */
```
Comment (marc-hb) [inline src/adsp_config.c:1594]: s/related/compatible/ ?

**Review by marc-hb — COMMENTED** [review]
The first commit has two copies of the same `/* parse "version" array elements */` code and computes manifest_version[] twice. It's removed in the second commit but the state between the two commits is weird/inconsistent. I guess it's just a git "fumble"?

File: src/adsp_config.c (hunk 2 above)
Comment (ktrzcinx) [reply src/adsp_config.c]: changed

File: src/adsp_config.c (hunk 1 above)
Comment (ktrzcinx) [reply src/adsp_config.c]: Ok, so I extracted `find_config_parser()` function, then error flow is everywhere inside `if` statement and success flow is outside `if`

**Comment by ktrzcinx** [comment]
Checkpatch failed, but actually I don't know why, perhaps some checkpatch issue?

**Comment by marc-hb** [comment]
> Checkpatch failed, but actually I don't know why, perhaps some checkpatch issue?

I think so, same bug I just described in https://github.com/thesofproject/rimage/pull/29#issuecomment-710748554

Looks like this bug is going to trigger every time there is a `struct` :-( Maybe it has been fixed upstream?

File: src/adsp_config.c [hunk 3]
```diff
@@ -1532,6 +1532,33 @@ static int parse_adsp_config_v2_5(const toml_table_t *toml, struct adsp *out,
 	return 0;
 }
 
+static int parse_version(toml_table_t *toml, int64_t version[2])
```
Comment (lgirdwood) [inline src/adsp_config.c:1535]: What constitutes a correct version string here, can you say in an inline comment.

File: src/adsp_config.c [hunk 4]
```diff
@@ -1559,13 +1559,35 @@ static int parse_version(toml_table_t *toml, int64_t version[2])
 	return 0;
 }
 
-static inline bool check_config_version(int major, int minor, const int64_t *version)
+struct parser {
```
Comment (lgirdwood) [inline src/adsp_config.c:1562]: version_parser ? It seems specific for parsing versions only ?

**Review by lgirdwood — CHANGES_REQUESTED** [review]
Just some more comments needed.

File: src/adsp_config.c (hunk 4 above)
Comment (ktrzcinx) [reply src/adsp_config.c:1562]: no, it's parser for whole configuration file, version is used only to select right parser

File: src/adsp_config.c (hunk 3 above)
Comment (ktrzcinx) [reply src/adsp_config.c:1538]: version is not stored in string, it is in toml array of integers. I've added this information as comment.

File: src/adsp_config.c (hunk 4 above)
Comment (ktrzcinx) [reply src/adsp_config.c]: renamed to `config_parser`

**Review by lgirdwood — APPROVED** [review]

**Comment by lgirdwood** [comment]
CI showing checkpatch errors.

**Pull request merged**